pages/index.js

```diff
@@ -1,31 +1,109 @@
-	let cacheClient;
-
-
-	useEffect(() => {
-		async function setupMomento() {
-			initializeTopicClient();
-			initializeCacheClient();
-			getRoomList();
-		}
-
-		if (credentials && !topicClient) {
-			setupMomento();
-		}
-	}, [credentials]);
-
-	const getAuthToken = async () => {
+  const [topicClient, setTopicClient] = useState(null);
+  const [cacheClient, setCacheClient] = useState(null);
+  const [chatRooms, setChatRooms] = useState([]);
+  const [credentials, setCredentials] = useState(null);
+
+  const login = useCallback(async () => {
+    const username =
+      `${faker.color.human()}-${faker.animal.type()}`.toLowerCase();
+    const response = await fetch(
+      `${process.env.NEXT_PUBLIC_AUTH_BASE_URL}/authenticate`,
+      {
+        method: "POST",
+        body: JSON.stringify({
+          username,
+        }),
+      }
+    );
+
+    const token = await response.json();
+    const userInfoResponse = await fetch(
+      `${process.env.NEXT_PUBLIC_AUTH_BASE_URL}/userinfo`
+    );
+
+    const userInfo = await userInfoResponse.json();
+    const credentials = {
+      auth: token,
+      user: { ...userInfo, username },
+    };
+
+    sessionStorage.setItem("credentials", JSON.stringify(credentials));
+    setCredentials(credentials);
+  }, []);
+
+  useEffect(() => {
+    const storedCredentials = sessionStorage.getItem("credentials");
+    if (storedCredentials) {
+      const creds = JSON.parse(storedCredentials);
+      if (
+        !creds.user?.claims?.momento?.exp ||
+        creds.user?.claims?.momento?.exp < Date.now()
+      ) {
+        sessionStorage.removeItem("credentials");
+        login();
+      } else {
+        setCredentials(creds);
+      }
+    } else {
+      login();
+    }
+  }, [login]);
+
+  const getRoomList = useCallback(async () => {
+    const roomListResponse = await cacheClient?.setFetch(
+      "chat",
+      "chat-room-list"
+    );
+    if (roomListResponse instanceof CacheSetFetch.Hit) {
+      setChatRooms(roomListResponse.valueArrayString().sort());
+    } else {
+      setChatRooms([]);
+    }
+  }, [cacheClient]);
+
+  const initializeCacheClient = useCallback(() => {
+    if (!cacheClient) {
+      const client = new CacheClient({
+        configuration: Configurations.Browser.v1(),
+        credentialProvider: CredentialProvider.fromString({
+          authToken: credentials.user.claims.momento.token,
+        }),
+        defaultTtlSeconds: 3600,
+      });
+
+      setCacheClient(client);
+    }
+  }, [cacheClient, credentials]);
+
+  const initializeTopicClient = useCallback(() => {
+    if (!topicClient) {
+      const client = new TopicClient({
+        configuration: Configurations.Browser.v1(),
+        credentialProvider: CredentialProvider.fromString({
+          authToken: credentials.user.claims.momento.token,
+        }),
+      });
+
+      setTopicClient(client);
+    }
+  }, [topicClient, credentials]);
+
+	const getAuthToken = useCallback(async () => {
```

Comment:
Although it was an unused function, I think it is in the process of being implemented in some way, so I kept it in place instead of being deleted.